### PR TITLE
gh-118335: Configure Tier 2 interpreter at build time

### PR DESCRIPTION
**Pull request by gvanrossum**
The code for Tier 2 is now only compiled when configured with `--enable-experimental-jit[=yes|interpreter]`. We drop support for `PYTHON_UOPS` and -`Xuops`, but you can disable the interpreter or JIT at runtime by setting `PYTHON_JIT=0`. You can also build it without enabling it by default using `--enable-experimental-jit=yes-off`; enable with `PYTHON_JIT=1`.

@mdboom See note in issue about benchmarking setup.

~~@markshannon An alternative approach would be to keep the APIs but make them always fail. In particular, there are some `PyUnstable` APIs that will only exist when `_Py_TIER2` is defined. Thoughts?~~

~~**EDIT:** In particular, we could arguably keep offering `PyUnstable_{Get,Set}Optimizer`, so 3rd parties could create and activate their own optimizers, like in your original design (which I cannot find back -- you had something where 3rd parties could register optimizations and parameters indicating the cost and benefit, or something). However, the rest of the optimizer machinery is now dependent on executors (starting with `_PyOptimizer_Optimize`, ). If we wanted to keep supporting both optimizers and executors, we'd have to do something different in `ENTER_EXECUTOR` too. (Possibly in both cases the key issue is that `GOTO_TIER_TWO(executor)` assumes JIT or Tier 2 IR.)~~

~~**EDIT 2:** Another consideration is whether we'd still want the runtime `-Xuops` flag and the `PYTHON_UOPS` env var, when enabled at build time. Or perhaps at least `PYTHON_UOPS=0` to *disable* at runtime (which is a simple change in pylifecycle.c). I'd use that when debugging tests with `PYTHON_LLTRACE=N`, since the debug output from the test infrastructure might overwhelm that for the test proper.~~

~~Yet another option would be to leave everything as it was, *except* change the default with the build-time flag.~~

* Issue: gh-118335

----
📚 Documentation preview 📚: https://cpython-previews--118339.org.readthedocs.build/

### CODE REVIEW

**Comment by gvanrossum** [comment]
WASI build is [failing](https://github.com/python/cpython/actions/runs/8855844443/job/24321353589?pr=118339), just like it did [before](https://github.com/python/cpython/pull/117908) when I had an option to *disable* Tier 2.

I'm away from my Windows laptop so I can't debug the Windows failures yet.

**Comment by gvanrossum** [comment]
There's also this failure (a hang in test_multiprocessing, apparently): https://github.com/python/cpython/actions/runs/8857124270/job/24324215355?pr=118339

PS. I've updated the discussion in the initial comment.

**Comment by gvanrossum** [comment]
[Benchmark](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240426-3.13.0a6+-8cb71dd) says 1% faster (in Tier 1 mode).

**Comment by gvanrossum** [comment]
Looks like Windows JIT builds now pass. WASI build still fails. I wonder if there's something in the WASI build similar to the Windows problem (which had to do with `_Py_TIER2` never getting set in a place where it was tested).

**Review by mdboom — CHANGES_REQUESTED** [review]
This may already be on your radar, but a reminder to also update the Windows `build.bat` script to support this 4-way option.

**Comment by mdboom** [comment]
I have a draft PR to support this in the benchmarking infrastructure.  I don't think it's possible to support the old way and the new way simultaneously, so we'll want to merge these around the same moment.

**Comment by gvanrossum** [comment]
> This may already be on your radar, but a reminder to also update the Windows `build.bat` script to support this 4-way option.

Thanks for the reminder. For now I'm only going to allow JIT on/off there, I have to think about the right way to select the default on/off and interpreter instead of JIT. (Maybe it needs a separate property.)

**Comment by gvanrossum** [comment]
> That's probably fine for now, but it will mean we can't benchmark Tier 2 on Windows. But maybe @markshannon is the only person who cares, and even then, we aren't optimizing Tier 2 for speed.

Okay, how about this plan. We merge this PR once I get the WASI build going (I'm now just disabling the offending test on WASI in debug mode), and then I can see if I have time to come up with a build.bat flag to enable the Tier 2 interpreter in a separate PR. That could even happen post beta 1, as could the WASI fix (if we care).

**Comment by gvanrossum** [comment]
For the WASI crash I've filed gh-118445.
For the extra Windows build options I've filed gh-118446.

**Comment by gvanrossum** [comment]
WASI bliss at last.

@brandtbucher Do you have time to review this?

@mdboom I peeked at your [new-flags](https://github.com/faster-cpython/benchmarking/compare/main...new-flags) branch in the benchmarking repo. We lose the ability to benchmark Tier 2 on Windows (for now) -- is that reflected in the diff there? (As an only occasional user of the UI there I rely on the script to let me know if I've selected an impossible combination.)

**Comment by brandtbucher** [comment]
> @brandtbucher Do you have time to review this?

Yep, looking at it now.

File: Doc/whatsnew/3.13.rst [hunk 1]
```diff
@@ -905,19 +905,35 @@ The internal architecture is roughly as follows.
...
 
+The ``--enable-experimental-jit`` flag has the following optional values:
+
+* ``no`` (default) -- Disable the entire Tier 2 and JIT pipeline.
+
+* ``yes`` (default if the flag is present without optional value)
+  -- Enable the JIT. To disable the JIT at runtime,
+  pass the environment variable ``PYTHON_JIT=0``.
+
+* ``yes-default-off`` -- Build the JIT but disable it by default.
```
Comment (brandtbucher) [inline Doc/whatsnew/3.13.rst]: Bikeshedding: I'd personally word `yes-default-off` and `interpreter-default-off` without the "default" part. So just `yes-off` and `interpreter-off`.

File: Doc/whatsnew/3.13.rst [hunk 2]
```diff
@@ -905,19 +905,35 @@ The internal architecture is roughly as follows.
...
+  To emable the JIT at runtime, pass the environment variable
+  ``PYTHON_JIT=1``.
+
+* ``interpreter`` -- Enable the Tier 2 interpreter but disable the JIT.
+  The interpreter can be disabled by running with
+  ``PYTHON_JIT=0``.
+
+(On Windows, use ``PCbuild/build.bat --enable-jit`` to enable the JIT.)
```
Comment (brandtbucher) [inline Doc/whatsnew/3.13.rst:935]: We should probably have the same options exposed in `PCbuild`, right? Doesn't need to be in this PR, though.

File: Doc/whatsnew/3.13.rst [hunk 3]
```diff
@@ -905,19 +905,35 @@ The internal architecture is roughly as follows.
...
+The ``--enable-experimental-jit`` flag has the following optional values:
+
+* ``no`` (default) -- Disable the entire Tier 2 and JIT pipeline.
+
+* ``yes`` (default if the flag is present without optional value)
+  -- Enable the JIT. To disable the JIT at runtime,
+  pass the environment variable ``PYTHON_JIT=0``.
+
+* ``yes-default-off`` -- Build the JIT but disable it by default.
+  To emable the JIT at runtime, pass the environment variable
```
Comment (brandtbucher) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  To enable the JIT at runtime, pass the environment variable
```

File: Lib/test/support/__init__.py [hunk 4]
```diff
@@ -2540,6 +2540,7 @@ def exceeds_recursion_limit():
 def without_optimizer(func):
     try:
         import _testinternalcapi
+        from _testinternalcapi import get_optimizer, set_optimizer
```
Comment (brandtbucher) [inline Lib/test/support/__init__.py]: This is sort of funky. Maybe remove the old import:

```suggestion
        from _testinternalcapi import get_optimizer, set_optimizer
```

...and use `get_optimizer` and `set_optimizer` directly below?

File: Misc/NEWS.d/next/Core and Builtins/2024-04-26-14-06-18.gh-issue-118335.SRFsxO.rst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Change how to use the tier 2 interpreter. Instead of running Python with
+``-X uops`` or setting the environment variable ``PYTHON_UOPS=1``, this
+choice is now made at build time by configuring with
+``--enable-experimental-tier2``.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-26-14-06-18.gh-issue-118335.SRFsxO.rst]: This should be updated.

File: PCbuild/_testinternalcapi.vcxproj [hunk 6]
```diff
@@ -108,6 +108,12 @@
       <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
     </ProjectReference>
   </ItemGroup>
+  <ItemDefinitionGroup>
+    <ClCompile>
+      <PreprocessorDefinitions Condition="'$(UseJIT)' == 'true'">_Py_JIT;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions Condition="'$(UseJIT)' == 'true'">_Py_TIER2;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (brandtbucher) [inline PCbuild/_testinternalcapi.vcxproj]: I *think* these can be combined:

```suggestion
      <PreprocessorDefinitions Condition="'$(UseJIT)' == 'true'">_Py_JIT;_Py_TIER2;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```

File: PCbuild/pythoncore.vcxproj [hunk 7]
```diff
@@ -105,6 +105,7 @@
       <PreprocessorDefinitions>_USRDLL;Py_BUILD_CORE;Py_BUILD_CORE_BUILTIN;Py_ENABLE_SHARED;MS_DLL_ID="$(SysWinVer)";%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <PreprocessorDefinitions Condition="$(IncludeExternals)">_Py_HAVE_ZLIB;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <PreprocessorDefinitions Condition="'$(UseJIT)' == 'true'">_Py_JIT;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions Condition="'$(UseJIT)' == 'true'">_Py_TIER2;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (brandtbucher) [inline PCbuild/pythoncore.vcxproj]: Ditto:

```suggestion
      <PreprocessorDefinitions Condition="'$(UseJIT)' == 'true'">_Py_JIT;_Py_TIER2;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```

File: Python/bytecodes.c [hunk 8]
```diff
@@ -2361,6 +2361,7 @@ dummy_func(
             CHECK_EVAL_BREAKER();
             assert(oparg <= INSTR_OFFSET());
             JUMPBY(-oparg);
+            #ifdef _Py_TIER2
             #if ENABLE_SPECIALIZATION
```
Comment (brandtbucher) [inline Python/bytecodes.c:2365]: I personally find this form a bit easier to reason about than the nesting:

```suggestion
            #if defined(_Py_TIER2) && ENABLE_SPECIALIZATION
```

File: Python/bytecodes.c [hunk 9]
```diff
@@ -2386,6 +2387,7 @@ dummy_func(
                 ADVANCE_ADAPTIVE_COUNTER(this_instr[1].counter);
             }
             #endif  /* ENABLE_SPECIALIZATION */
+            #endif /* _Py_TIER2 */
```
Comment (brandtbucher) [inline Python/bytecodes.c:2390]: ...if the suggestion above is applied.

```suggestion
```

File: configure.ac [hunk 10]
```diff
@@ -1768,14 +1768,23 @@ fi
...
               [],
               [enable_experimental_jit=no])
+case $enable_experimental_jit in
+  no) enable_experimental_jit=no ;;
+  yes) enable_experimental_jit="-D_Py_JIT -D_Py_TIER2=1" ;;
+  yes-default-off) enable_experimental_jit="-D_Py_JIT -D_Py_TIER2=3" ;;
+  interpreter) enable_experimental_jit="-D_Py_TIER2=4" ;;
+  interpreter-default-off) enable_experimental_jit="-D_Py_TIER2=6" ;;  # Secret option
+  *) AC_MSG_ERROR(
+      [invalid argument: --enable-experimental-jit=$enable_experimental_jit; expected no|yes|yes-default-off|interpreter]) ;;
```
Comment (brandtbucher) [inline configure.ac]: I find it a bit hard to reason about the relationship between `_Py_JIT` and the `_Py_TIER2` flags... I think there's redundant information here.

Maybe it would be clearer to just use `_Py_JIT` and have the different options be encoded in its bits. *Or* have orthogonal defines and get rid of the bit masks entirely:

- `_Py_JIT`: Defined for `yes`, `yes-off`.
- `_Py_JIT_ENABLED`: Defined for `yes` and `interpreter`.
- `_Py_JIT_INTERPRETER`: Defined for `interpreter` and `interpreter-off`.

(I *think* these make sense, in terms of usefulness?)

Again, this stuff is all internal, so we can tweak it if need be.

File: Tools/jit/README.md [hunk 11]
```diff
@@ -7,18 +7,18 @@ This version of CPython can be built with an experimental just-in-time compiler.
 
 The JIT compiler does not require end users to install any third-party dependencies, but part of it must be *built* using LLVM[^why-llvm]. You are *not* required to build the rest of CPython using LLVM, or even the same version of LLVM (in fact, this is uncommon).
 
-LLVM version 16 is required. Both `clang` and `llvm-readobj` need to be installed and discoverable (version suffixes, like `clang-16`, are okay). It's highly recommended that you also have `llvm-objdump` available, since this allows the build script to dump human-readable assembly for the generated code.
```
Comment (brandtbucher) [inline Tools/jit/README.md:10]: Thanks for catching these!

File: Python/pylifecycle.c [hunk 12]
```diff
@@ -1272,30 +1274,30 @@ init_interp_main(PyThreadState *tstate)
     }
 
     // Turn on experimental tier 2 (uops-based) optimizer
+    // This is also needed when the JIT is enabled
+#ifdef _Py_TIER2
     if (is_main_interp) {
```
Comment (brandtbucher) [inline Python/pylifecycle.c:1279]: I know this briefly came up in the meeting, but why are we only doing this in the main interpreter again? (Doesn't necessarily need to be fixed here, just remembered and hot curious.)

File: Python/instrumentation.c [hunk 13]
```diff
@@ -1702,10 +1702,12 @@ instrument_lock_held(PyCodeObject *code, PyInterpreterState *interp)
         );
         return 0;
     }
+#ifdef _Py_TIER2
```
Comment (brandtbucher) [inline Python/instrumentation.c:1705]: Reading through this, I'm constantly torn between doing what you've done here or just making all of these APIs no-ops when tier two isn't built. I think the way you've done it is probably better, but all of these `#ifdef`s in mostly-unrelated files make me a tiny bit sad.

File: Python/bytecodes.c [hunk 14]
```diff
@@ -2416,6 +2419,9 @@ dummy_func(
             tstate->previous_executor = Py_None;
             Py_INCREF(executor);
             GOTO_TIER_TWO(executor);
+            #else
+            assert(0);
```
Comment (brandtbucher) [inline Python/bytecodes.c]: This should probably be a `Py_FatalError` instead of effectively a `NOP` on `NDEBUG` builds. ;)

File: Modules/_opcode.c [hunk 15]
```diff
@@ -367,7 +367,13 @@ _opcode_get_executor_impl(PyObject *module, PyObject *code, int offset)
                      Py_TYPE(code)->tp_name);
         return NULL;
     }
+#ifdef _Py_TIER2
     return (PyObject *)PyUnstable_GetExecutor((PyCodeObject *)code, offset);
+#else
+    PyErr_Format(PyExc_RuntimeError,
+                 "Executors is not available in this build");
```
Comment (brandtbucher) [inline Modules/_opcode.c]:
```suggestion
                 "Executors are not available in this build");
```

**Review by brandtbucher — COMMENTED** [review]
Thanks for taking the time to tackle this. Performing the surgery after-the-fact definitely seems like a tricky task.

File: PCbuild/_testinternalcapi.vcxproj (hunk 6 above)
Comment (gvanrossum) [reply PCbuild/_testinternalcapi.vcxproj]: Yes, they can, but I won't, in anticipation of changing the second one to `UseTIER2` to enable the Tier 2 interpreter on Windows. (I've got this half working. :-)

File: Python/bytecodes.c (hunk 8 above)
Comment (gvanrossum) [reply Python/bytecodes.c:2365]: Let's agree to disagree -- I like that all the uses of `ENABLE_SPECIALIZATION` look the same.

File: configure.ac (hunk 10 above)
Comment (gvanrossum) [reply configure.ac]: I tried that and something broke (a segfault in `_build_python`) and I couldn't figure out what I'd done wrong. All the JIT stuff that you're responsible for keeps using `#ifdef _Py_JIT`. All the stuff that is shared between the JIT and the Tier 2 interpreter checks for `_Py_TIER2` -- almost all in the form of `#ifdef _Py_TIER2`. There are only 2 places that have a different form of test, one in ceval.c (L758, where some variables only exist for the interpreter) and one in pylifecycle.c (checking `_Py_TIER2 & 2` to determine if the default should be on or off).

File: Python/pylifecycle.c (hunk 12 above)
Comment (gvanrossum) [reply Python/pylifecycle.c:1279]: I have no idea any more. I'll see if deleting this check causes any problems.

File: Python/instrumentation.c (hunk 13 above)
Comment (gvanrossum) [reply Python/instrumentation.c:1705]: Totally. What I did is easier, because otherwise I'd have to selectively keep certain parts of optimizer.c where the functionality is implemented. That would be a mess of functions surrounded by `#endif ... #ifdef`, or I'd have to move everything to the end of the file. I tend to prefer the shortest diff.

File: Python/bytecodes.c (hunk 14 above)
Comment (gvanrossum) [reply Python/bytecodes.c]: Interestingly, adding that causes `ENTER_EXECUTOR` to acquire the `HAS_ESCAPES_FLAG` flag. Fixing that (by updating `NON_ESCAPING_FUNCTIONS` in Tools/cases_generator/analyzer.py) causes `_FATAL_ERROR` to lose that flag. I'm going to assume that doesn't matter, because `Py_FatalError` never returns.

Comment (brandtbucher) [reply Python/bytecodes.c]: ...now I'm twisting my mind up in knots trying to figure out if `ENTER_EXECUTOR` *should* be considered escaping. But I don't think so (and I don't think it matters very much).

File: Python/pylifecycle.c (hunk 12 above)
Comment (gvanrossum) [reply Python/pylifecycle.c:1279]: Alas, deleting it does cause problems -- `test.test_capi.test_misc.TestPendingCalls.test_isolated_subinterpreter` hangs. So I'm gonna stick with what works.

File: configure.ac (hunk 10 above)
Comment (gvanrossum) [reply configure.ac]: Actually, what I tried was folding the bitmask into a single `_Py_JIT` flag, and that definitely was no improvement. I am trying to limit the number of predefined variables; apparently 1 didn't work, and I like how it turned out with 2 (JIT and TIER2), so I'm sticking with what I've got.

As you say, this is all internal and nobody else should care, so it should be easy to change later.

**Comment by gvanrossum** [comment]
@brandtbucher I believe I've done everything from your review that I gave a thumbs up, and for everything else I stated a reason.

@mdboom -- At Brandt's suggestion I've shortened the "default off" flag to `--enable-experimental-jit=yes-off`. I *think* this has no repercussions for your new-flags branch of the benchmarking repo, but you may want to double-check.

**Review by brandtbucher — APPROVED** [review]
> @brandtbucher I believe I've done everything from your review that I gave a thumbs up, and for everything else I stated a reason.

Yep!

**Comment by gvanrossum** [comment]
Okay, I've got the Windows flag stuff sorted out. It's a little different from Unix, because you apparently can't say "--experimental-jit=yes-off" while also supporting "--experimental-jit"; it treats the "=" as a space. Rather than becoming a .BAT file CLI parsing expert I chose the following option names:
- `--experimental-jit`
- `--experimental-jit-off` (build JIT, but default off)
- `--experimental-interpreter`
- `--experimental-interpreter-off` (undocumented, for Mark :-)
